### DISCUSSION TITLE
Delta Measures on Slicing Hierarchies

### DISCUSSION THREAD

**Discussion by shahin-y** [atoti/atoti#732] [Q&A] [answered]
Hi everyone,

This question is regarding [slicing hierarchies](https://docs.atoti.io/latest/lib/atoti/atoti.hierarchy.html#atoti.Hierarchy.slicing).

According to the documentation
> A slicing hierarchy is not aggregable at the top level, meaning that it does not make sense to aggregate data across all members of the hierarchy.
For instance, for an `As of date` hierarchy giving the current bank account `Balance` for a given date, it does provide (sic!) any meaningful information to aggregate the `Balance` across all the dates.

However, even in this example, it may make sense to show the difference of the account `Balance` between different `As of date` values.
For instance, if there are multiple reference dates that we would like to refer to by a symbol (e.g. `Yesterday` may refer to the last business day or any other arbitrary date), the resulting data model might looks like this:
![image](https://user-images.githubusercontent.com/12223752/210771119-c65a2f6e-f6dd-46e0-9563-de9264a26969.png)
with the following sample data:
|DaySymbol|Date|
|---|---|
|Today|2022-12-24|
|Yesterday|2022-12-23|
|Last EoM|2022-11-30|

If the hierarchy would not be slicing, filtering on the `DaySymbol` could be used to create a measure returning the difference of `Today`'s and `Yesterday`'s account `Balance`:
```python
m["Today Balance.SUM"] = tt.filter(tt.agg.sum(session.tables["Bank Accounts"]["Balance"]), cube.levels["DaySymbol"] == "Today")
m["Yesterday Balance.SUM"] = tt.filter(tt.agg.sum(session.tables["Bank Accounts"]["Balance"]), cube.levels["DaySymbol"] == "Yesterday")
m["Balance.SUM Delta (Today - Yesterday)"] = m["Today Balance.SUM"] - m["Yesterday Balance.SUM"]
```
Now if the hierarchy is slicing this doesn't work - and ideally it should remain slicing.

With a `date_shift()` on the other hand we could get the `Balance` for another date even if the hierarchy is slicing. Unfortunately,  the offset cannot be dynamic (see #725). One way to deal with that would be to derive the new offset every time the `SymbolicDates` data changes and update/overwrite all related measures.

However, is there maybe a better approach to accomplish this?

**Comment by rcolle** [comment]
Hi!
Have you considered using [tt.at](https://docs.atoti.io/latest/lib/atoti/atoti.function.at.html#atoti.at) instead of `filter` to create the today/yesterday measures? This should do the trick.

**Reply by shahin-y** [reply]
Hi,

Thank you for your prompt reply!

Do you mean using `tt.at` like this?
```python
m["at Today Balance.SUM"] = tt.at(tt.agg.sum(session.tables["Bank Accounts"]["Balance"]), cube.levels["DaySymbol"] == "Today")
m["at Yesterday Balance.SUM"] = tt.at(tt.agg.sum(session.tables["Bank Accounts"]["Balance"]), cube.levels["DaySymbol"] == "Yesterday")
m["at Balance.SUM Delta (Today - Yesterday)"] = m["at Today Balance.SUM"] - m["at Yesterday Balance.SUM"]
```
Like this the `DaySymbol` has to be added and then only the one resolving to the `Date` set in the default filter is displayed:
![image](https://user-images.githubusercontent.com/12223752/210842180-4951bc2a-2eb9-4904-ab86-dc7fafe7e98f.png)
Slicing on the `Date` gives us values for the other measures but not really what we were trying to achieve:
![image](https://user-images.githubusercontent.com/12223752/210842865-20130147-e043-45d8-92c2-2671318eebd6.png)
If the `Date` hierarchy is not slicing it works nicely though:
![image](https://user-images.githubusercontent.com/12223752/210843395-720944a3-1c0d-46e1-a97c-89180f224ed9.png)

It seems like the default filter is taking precedence or am I missing something regarding the `tt.at` function?

For the sake of completeness, please find the example code below.
<details>
  <summary>Example</summary>

  ```python
import atoti as tt
import pandas as pd
from datetime import datetime

authentication=tt.BasicAuthenticationConfig()

session = tt.Session(
    port=9101
)

base = session.create_table("Bank Accounts",
    keys=["Date", "ID"],
    types={"Date" : tt.type.LOCAL_DATE, "ID": tt.type.STRING, "Balance": tt.type.DOUBLE, "Continent" : tt.type.STRING, "Country" : tt.type.STRING})

symbolic_dates = session.create_table("SymbolicDates",
                                      keys=["DaySymbol", "Date"],
                                      types={"DaySymbol": tt.type.STRING, "Date" : tt.type.LOCAL_DATE,})

tt.Table.join(base, symbolic_dates, mapping={"Date": "Date"})

cube = session.create_cube(session.tables["Bank Accounts"], name="demo cube", mode="auto")
m = cube.measures

cube.hierarchies["Date"].slicing = True
cube.levels["Date"].order = tt.NaturalOrder(ascending=False)
cube.hierarchies["location"] = [cube.levels["Continent"], cube.levels["Country"]]
# cube.schema
df_bank_accounts = pd.DataFrame({
    "Date" : [datetime.strptime("2022-12-23", "%Y-%m-%d"), datetime.strptime("2022-12-23", "%Y-%m-%d"), datetime.strptime("2022-12-23", "%Y-%m-%d"), datetime.strptime("2022-12-24", "%Y-%m-%d"), datetime.strptime("2022-12-24", "%Y-%m-%d"), datetime.strptime("2022-12-24", "%Y-%m-%d")],
    "ID": ["1", "2", "3", "1", "2", "3"],
    "Balance" : [1, 2, 3, 10,20, 30,],
    "Continent": ["EU", "EU", "EU", "EU", "EU", "EU"],
    "Country": ["DE", "FR", "DE", "DE", "FR", "DE"]
})

df_symbolic_date = pd.DataFrame({
    "DaySymbol": ["Today", "Yesterday", "Last EoM"],
    "Date" : [datetime.strptime("2022-12-24", "%Y-%m-%d"), datetime.strptime("2022-12-23", "%Y-%m-%d"), datetime.strptime("2022-11-30", "%Y-%m-%d"),],
})
session.tables["Bank Accounts"].load_pandas(df_bank_accounts)
session.tables["SymbolicDates"].load_pandas(df_symbolic_date)

m["Today Balance.SUM"] = tt.filter(tt.agg.sum(session.tables["Bank Accounts"]["Balance"]), cube.levels["DaySymbol"] == "Today")
m["Yesterday Balance.SUM"] = tt.filter(tt.agg.sum(session.tables["Bank Accounts"]["Balance"]), cube.levels["DaySymbol"] == "Yesterday")

m["Balance.SUM Delta (Today - Yesterday)"] = m["Today Balance.SUM"] - m["Yesterday Balance.SUM"]
# Calculated measure: ([Measures].[Today Balance.SUM] - [Measures].[Yesterday Balance.SUM])

m["at Today Balance.SUM"] = tt.at(tt.agg.sum(session.tables["Bank Accounts"]["Balance"]), cube.levels["DaySymbol"] == "Today")
m["at Yesterday Balance.SUM"] = tt.at(tt.agg.sum(session.tables["Bank Accounts"]["Balance"]), cube.levels["DaySymbol"] == "Yesterday")
m["at Balance.SUM Delta (Today - Yesterday)"] = m["at Today Balance.SUM"] - m["at Yesterday Balance.SUM"]
  ```
</details>

**Reply by rcolle** [reply]
I see. I read quicky through the question and didn't fully get the problem the first time.
Let me try to think a bit more about it and come back to you.

**Reply by tibdex** [reply]
Hi Shahin, this feature request is being tracked in an [internal ticket](https://activeviam.atlassian.net/browse/PIVOT-6956). We'll keep you updated here when progress is made.

**Reply by shahin-y** [reply]
Hi Thibault, Thanks for the update!

**Comment by rcolle** [comment, marked as answer]
Hi @shahin-y ,

Have you considered "hard-coding" the dates in your measures and updating them when the "symbol" changes?

You would:
- Not load the "Symbolic dates" table as a store in Atoti
- Create your measures with hard-coded dates, e.g.
```python
m["at Today Balance.SUM"] = tt.at(tt.agg.sum(session.tables["Bank Accounts"]["Balance"]), cube.levels["Date"] == "2023-03-27")
m["at Yesterday Balance.SUM"] = tt.at(tt.agg.sum(session.tables["Bank Accounts"]["Balance"]), cube.levels["Dates"] == "2023-03-24")
m["Balance.SUM Delta (Today - Yesterday)"] = m["at Today Balance.SUM"] - m["at Yesterday Balance.SUM"]
```
- change the "at" measures whevener the symbolic date change, e.g. tomorrow execute the following code:
```python
m["at Today Balance.SUM"] = tt.at(tt.agg.sum(session.tables["Bank Accounts"]["Balance"]), cube.levels["Date"] == "2023-03-28")
m["at Yesterday Balance.SUM"] = tt.at(tt.agg.sum(session.tables["Bank Accounts"]["Balance"]), cube.levels["Dates"] == "2023-03-27")
```

Re-defining measures is an instantaneous operation with Atoti, which does not require to start/stop the application, so this shouldn't be a problem performance-wise.

**Reply by rcolle** [reply]
Side note: in the real application, you would probably have the symbolic dates in a python dict, and a function to define all measures.
For instance
```python
symbolic_dates = {
  "TODAY": "2023-03-27",
  "YESTERDAY": "2023-03-24"
}

def create_at_measure(symbolic_dates):
  m["at Today Balance.SUM"] = tt.at(tt.agg.sum(session.tables["Bank Accounts"]["Balance"]), cube.levels["Date"] == symbolic_dates["TODAY"])
  m["at Yesterday Balance.SUM"] = tt.at(tt.agg.sum(session.tables["Bank Accounts"]["Balance"]), cube.levels["Dates"] == symbolic_dates["YESTERDAY"])
```

And call `create_at_measures` whenever you update `symbolic_dates`

**Reply by shahin-y** [reply]
Hi @rcolle,

Thank you very much for your suggestion! This approach works quite well and that way measures could be updated dynamically.

Unfortunately, it seems that re-defining the measures while the data stores are updated at the same time can lead to some unexpected behavior under heavy loads, i.e. internal errors and overall performance hits. I have created an [internal ticket](https://activeviam.atlassian.net/browse/UKS-5628) for this as this is out of scope for this discussion. In smaller projects or within notebooks this should be of no concern though.
